src/commands/import-resolver.ts

```diff
@@ -98,3 +119,95 @@ export function createImportResolver({
     return false;
   };
 }
+
+export interface ExtensionMap {
+  input: Record<string, string[]>;
+  output: Record<string, string[]>;
+}
+
+/** map plugin inputs & outputs to make lookups easier */
+export function mapPluginExtensions(plugins: SnowpackPlugin[]): ExtensionMap {
+  const extMap: ExtensionMap = {input: {}, output: {}};
+  plugins.forEach((plugin) => {
+    const inputs = Array.isArray(plugin.input) ? plugin.input : [plugin.input];
+    const outputs = Array.isArray(plugin.output) ? plugin.output : [plugin.output];
+
+    // given a file input ('.svelte'), what will the output extensions be? (['.js', '.css'])
+    inputs.forEach((ext) => {
+      if (!extMap.input[ext]) extMap.input[ext] = [];
+      extMap.input[ext] = [...new Set([...extMap.input[ext], ...outputs])]; // only keep unique extensions
+    });
+
+    // given a file output ('.css'), what could the input extension be? (['.css', '.scss', '.svelte'])
+    outputs.forEach((ext) => {
+      if (!extMap.output[ext]) extMap.output[ext] = [ext]; // an output must always possibly come from itself (not true for inputs)
+      extMap.output[ext] = [...new Set([...extMap.output[ext], ...inputs])];
+    });
+  });
+
+  return extMap;
+}
+
+/** given an absolute URL, which file on disk does this point to? */
+export function urlToFile(
```

Comment:
Focus 4: These 2 functions—`urlToFile()` and `fileToURLs()`—were important helpers. For myself, when going between `dev`, `build`, and `install`, it can become easily-confusing.

I did find clarity when I separated the logic between source file -> build file, and URL request -> source file. But in order to think this way, we have to develop some data structure that’s easily-traversable from the build pipeline, based on extensions.

While I think these 2 functions are really clear helpers in name and purpose, the implementation itself (including that crazy `mapPluginExtensions()` function) can be improved. I’d be sad to see these disappear, but they could totally be rewritten if needed.